### PR TITLE
Simplify kernel stubs and declaration

### PR DESCRIPTION
**Pull request by upsj**
This PR uses a handful of macros to simplify the `common_kernels.hpp.inc` and `*_kernels.hpp` files to make it easier to add new kernels. The majority of the changes were done by text replacement.

### CODE REVIEW

File: core/device_hooks/common_kernels.inc.cpp [hunk 1]
```diff
@@ -79,44 +79,93 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #endif  // GKO_HOOK_MODULE
 
 
+#define GKO_STUB(_macro) _macro GKO_NOT_COMPILED(GKO_HOOK_MODULE)
```
Comment (tcojean) [inline core/device_hooks/common_kernels.inc.cpp:82]: As an include, we really should add `<ginkgo/core/base/types.hpp>` as well.

File: core/device_hooks/common_kernels.inc.cpp [hunk 2]
```diff
@@ -79,44 +79,93 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+
+#define GKO_STUB_VALUE_CONVERSION(_macro)                             \
+    template <typename SourceType, typename TargetType>               \
+    _macro(SourceType, TargetType) GKO_NOT_COMPILED(GKO_HOOK_MODULE); \
+    GKO_INSTANTIATE_FOR_EACH_VALUE_CONVERSION(_macro)
+
+#define GKO_STUB_NON_COMPLEX_VALUE_TYPE(_macro)          \
+    template <typename ValueType>                        \
+    _macro(ValueType) GKO_NOT_COMPILED(GKO_HOOK_MODULE); \
+    GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(_macro)
```
Comment (tcojean) [inline core/device_hooks/common_kernels.inc.cpp:92]:
```suggestion
#define GKO_STUB_NON_COMPLEX_VALUE_TYPE(_macro)          \
    template <typename ValueType>                        \
    _macro(ValueType) GKO_NOT_COMPILED(GKO_HOOK_MODULE); \
    GKO_INSTANTIATE_FOR_EACH_NON_COMPLEX_VALUE_TYPE(_macro)
```

**Review by tcojean — APPROVED** [review]
Pretty good code reduction.

The script `create_new_algorithm.sh` will need to be updated but that can be done later.

File: core/device_hooks/common_kernels.inc.cpp (hunk 2 above)
Comment (upsj) [reply core/device_hooks/common_kernels.inc.cpp]: Good catch, thanks!

**Review by pratikvn — APPROVED** [review]
LGTM! Will make rebasing a bit annoying again, but better to add this now rather than later :)

**Comment by upsj** [comment]
@pratikvn yeah sorry, I added a bunch of those PRs in the last few weeks. Hopefully, that should be it for major changes, though :)

**Comment by pratikvn** [comment]
@upsj, no worries :) I think it is always better to have these types of improvements earlier rather than later.

**Comment by upsj** [comment]
rebase!

**Pull request merged**